### ISSUE TITLE
Cases when Hackage information is not updated

### ISSUE THREAD

**Issue by Anton-Latukha** [repology/repology-updater#1064]
I noticed that the project reports wrong Hackage versions, for example for the most famous text formats converter `pandoc`:

Repology reports to us that [newest `pandoc`  on Hackage is `2.2.1`](https://repology.org/project/haskell:pandoc/versions), but how can that be when all Haskell ecosystem today releases everything solely through Hackage.

Really, [on Hackage the latest `pandoc` released is `2.9.2.1`]( https://hackage.haskell.org/package/pandoc)

`2.2.1` version is 2018-05-11, it has been two years of constant development from that point, `2.2.1 -> 2.9.2.1` means that by Haskell version policy terms there were 7 major breakage releases in these two years, Haskell packages mostly curated by their `xx.xx.**.**` first and second digits since those mean the API stable releases/branches.

---

I know there is packdeps.haskellers.com and `pacdeps` utility that more canonically determine real last versions. Webservice currently does not respect deprecated versions (like when developers released it as API breakage but then due to no API breackage in real - rolled down the release number into a minor release and deprecated that major release increase). I submitted a report on that, for web service to respect the deprecations. `pacdeps` has a `--preferred` flag to properly respect the deprecations, deprecations of the major release increases by developers should be respected due to otherwise the service would indicate the old placebo major release, when the development goes and latest releases are in the previous numbers.

I do not use `pacdeps` at the moment, so I do not know if it would be useful to your project, but I think most probably.

**Comment by AMDmi3** [comment]
That's what Repology uses:

```
% wget -qO- https://hackage.haskell.org/packages/index.tar.gz | tar -tf- | awk -F/ '$1=="pandoc"{print $2}' | sort -g | tail -10
2.0.3
2.0.4
2.0.5
2.0.6
2.1
2.1.1
2.1.2
2.1.3
2.2
2.2.1
```

Given than overall hackage freshness is still >99%, it could be a problem with that index file generation which affects some individual packages.

> currently does not respect deprecated versions

We don't either, see #838

**Comment by Anton-Latukha** [comment]
Ok, I would know about this and would look what I can contribute here.

Most probably there is a definite canonical Haskell tooling way or tool to directly resolve this information.

Maybe I even ask the most knowledgeable person and his club about this on the livestream in a couple of hours.

**Comment by Anton-Latukha** [comment]
Also this summer there is work happening on extending Hackage API, particularly for "badges", which in fact is API returning versioning info: https://summer.haskell.org/ideas.html#hackage-badges

**Comment by Anton-Latukha** [comment]
Currently I found:

As you probably know the Cabal is the Haskell official package manager and it directly works with Hackage.

#### Update Hackage local DB
```fish
cabal v2-update
```

Logically it does the same as `wget -qO- https://hackage.haskell.org/packages/index.tar.gz`.

#### Get the metadata information
```
cabal info pandoc

* pandoc           (programs and library)
    Synopsis:      Conversion between markup formats
    Versions available: 1.19.2.4, 2.7.3, 2.8.0.1, 2.8.1, 2.9, 2.9.1, 2.9.1.1,
                        2.9.2, 2.9.2.1 (and 133 others)
    Versions installed: [ Not installed ]
    Homepage:      https://pandoc.org
    Bug reports:   https://github.com/jgm/pandoc/issues
    Description:   Pandoc is a Haskell library for converting from one markup
                   format to another, and a command-line tool that uses this
                   library. It can read several dialects of Markdown and
                   (subsets of) HTML, reStructuredText, LaTeX, DocBook, JATS,
                   MediaWiki markup, DokuWiki markup, TWiki markup, TikiWiki
                   markup, Jira markup, Creole 1.0, Haddock markup, OPML, Emacs
                   Org-Mode, Emacs Muse, txt2tags, ipynb (Jupyter notebooks),
                   Vimwiki, Word Docx, ODT, EPUB, FictionBook2, roff man,
                   Textile, and CSV, and it can write Markdown,
                   reStructuredText, XHTML, HTML 5, LaTeX, ConTeXt, DocBook,
                   JATS, OPML, TEI, OpenDocument, ODT, Word docx, PowerPoint
                   pptx, RTF, MediaWiki, DokuWiki, XWiki, ZimWiki, Textile,
                   Jira, roff man, roff ms, plain text, Emacs Org-Mode,
                   AsciiDoc, Haddock markup, EPUB (v2 and v3), ipynb,
                   FictionBook2, InDesign ICML, Muse, LaTeX beamer slides, and
                   several kinds of HTML/JavaScript slide shows (S5, Slidy,
                   Slideous, DZSlides, reveal.js).
                   
                   In contrast to most existing tools for converting Markdown to
                   HTML, pandoc has a modular design: it consists of a set of
                   readers, which parse text in a given format and produce a
                   native representation of the document, and a set of writers,
                   which convert this native representation into a target
                   format. Thus, adding an input or output format requires only
                   adding a reader or writer.
    Category:      Text
    License:       GPL-2
    Author:        John MacFarlane <jgm@berkeley.edu>
    Maintainer:    John MacFarlane <jgm@berkeley.edu>
    Source repo:   git://github.com/jgm/pandoc.git
    Executables:   pandoc, trypandoc
    Flags:         static, embed_data_files, trypandoc
    Dependencies:  base-noprelude >=4.9 && <5, base-compat >=0.9,
                   syb >=0.1 && <0.8, containers >=0.4.2.1 && <0.7,
                   unordered-containers ==0.2.*, parsec ==3.1.*, mtl ==2.2.*,
                   exceptions >=0.8 && <0.11, filepath >=1.1 && <1.5,
                   process >=1.2.3 && <1.7, directory >=1.2.3 && <1.4,
                   bytestring >=0.9 && <0.11, text >=1.1.1.0 && <1.
...
```

That is how it reports version list, it has no additional keys get it separately or to prettify that.

There is a couple of old tools/libs relatively around this: `hackage-processing`, `Hayoo`, `package-description`

There is `hackage-db` that Nixpkgs relies on, that is the most solid, but it is a library. All roads keep leading to `peti`.

**Comment by Anton-Latukha** [comment]
So. We sorted it out.

---

For testing we would use `binary` package, it deprecated last two versions, `0.9`, `0.10` and so `0.8` is the real latest branch.

New `index` link is http://hackage.haskell.org/01-index.tar.gz
Previous is http://hackage.haskell.org/00-index.tar.gz

So looks like they also can further increase version someday.

It (01-index.tar.gz) has a new `pandoc` version and the whole shebang.

But that does not respect the deprecated versions, we can observe there is `binary` `0.9, 0.10`.

---

Also, `sort` has `--version-sort` sort option.

**Comment by Anton-Latukha** [comment]
Also `peti` has the `hackage-db` library that allows on top of it run `cabal -v0 v2-run show-package-versions binary` and it reports a clean list and respects the deprecations also, but this needs presetup and download/compilation of Haskell libraries, and better this tool be implemented separately by itself.

**Comment by Anton-Latukha** [comment]
---

Better to update local Hackage DB with Cabal, maybe share over the node instances in your setup. And then do it over official Haskell util - the `cabal`.

```
cabal info binary
...
    Versions available: 0.6.4.0, 0.7.5.0, 0.7.6.0, 0.7.6.1, 0.8.6.0, 0.8.7.0,
                        0.8.8.0, (0.9.0.0), (0.10.0.0) (and 37 others)
```

Deprecated versions are inside `(..)`

**Comment by Anton-Latukha** [comment]
So here we have.

If you would need a full list of all versions provided. Then - technically deprecated versions are legitimate, `cabal` just really avoids them to the end, and picks them only if there are no other options. And you can parse http://hackage.haskell.org/01-index.tar.gz.

Or if you want the list of only really used versions. You can find a `show-package-versions` in `haskell-db` project and run `cabal -v0 v2-run show-package-versions <package>` on that file - it would print all versions but not deprecations.

If you want to get the last version, or last versions on last several release branches - `cabal info <package>`

**Comment by AMDmi3** [comment]
Switched to `01-index.tar.gz`. Pandoc is fixed, hackage freshness rose to 99.9%.

Had to rewrite the parser to allow incremental tarfile in multiple passes. `preferred-versions` are collected, but not parsed and used yet. It shouldn't be too hard to implement parser for these and support deprecated versions, but that's not top priority right now.
